ddr/lib/ddr-scanner/dwarf/DwarfParser.cpp

```diff
@@ -79,9 +79,48 @@ dwarf_init(int fd,
 	 */
 	FILE *fp = NULL;
 	if (DW_DLV_OK == ret) {
-		stringstream ss;
-		ss << "dwarfdump " << filepath << " 2>&1";
-		fp = popen(ss.str().c_str(), "r");
+
+		/* On some Mac machines, calling "dwarfdump" in shell calls a version named
+		 * "llvm-dwarfdump" instead of the version of dwarfdump DDR uses, which is
+		 * instead named "dwarfdump-classic" on those systems. Thus, we first try to
+		 * find "dwarfdump-classic" directly in case "dwarfdump" is linked to
+		 * "llvm-dwarfdump". If we don't find "dwarfdump-classic", then the machine
+		 * we are on uses the correct version of dwarfdump.
+		 */
+		fp = popen("xcrun -f dwarfdump-classic 2>&1", "r");
+		if (NULL != fp) {
+			stringstream toolpath;
+			char *buffer = NULL;
+			size_t cap = 0;
+			ssize_t len = getline(&buffer, &cap, fp);
+			if ((-1 == len) || (buffer[0] != '/')) {
+				pclose(fp);
+				fp = NULL;
+				fp = popen("xcrun -f dwarfdump 2>&1", "r");
+				if (NULL == fp) {
+					ret = DW_DLV_ERROR;
+					setError(error, DW_DLE_IOF);
+				}
+				len = getline(&buffer, &cap, fp);
+				if ((-1 == len) || (buffer[0] != '/')) {
+					ret = DW_DLV_ERROR;
+					setError(error, DW_DLE_IOF);
+				}
+			}
+			/* remove the newline consumed by getline */
+			if ((len > 0) && ('\n' == buffer[len - 1])) {
+				buffer[len - 1] = '\0';
+			}
+			toolpath << buffer;
```

Comment:
`buffer` could be `NULL` here (if no dwarfdump is available).

Reply:
I don't believe so. From `man getline` on OSX:
```
RETURN VALUES
     The getdelim() and getline() functions return the number of characters written, excluding the terminating NUL character.  The value -1 is
     returned if an error occurs, or if end-of-file is reached.
```
So if `len != -1` this means that characters are successfully read into the buffer.